### PR TITLE
Enhancement create calandar invitation

### PR DESCRIPTION
**Pull request by willkoua**
| Q                                                      | R
| ------------------------------------------ | -------------------------------------------
| Type of contribution ?                      | [Enhancement]
| Tickets (_issues_) concerned               | [78]

---

##### What have you changed ?
I add feature for send invitation icalandar

##### How did you change it ?
...

##### Verification :

**PR standard**
 - [x] Branches naming convention: 
     - `prefix-snake_case_description`
     - ex: `enhancement-create_new_cell`
     - ex: `fix-pep8_standard_on_cell_model`
 - [x] Fill in this automatically generated PR template
 - [x] Do not include issue numbers in the PR title
 - [x] Include screenshots and animated GIFs in your pull request whenever possible

**Code standard**
 - [x] This Pull-Request fully meets the requirements defined in the issue
 - [x] Add or modify the attached tests
 - [x] Follow the Python Styleguide
 - [x] Document new code based on the Documentation Styleguide
 - [x] Use I18N functions when you add some text

### CODE REVIEW

File: source/apiVolontaria/volunteer/serializers.py [hunk 1]
```diff
@@ -414,11 +415,39 @@ class ParticipationBasicSerializer(serializers.ModelSerializer):
 
     # Explicitly declare the BooleanField to make it "required"
     standby = serializers.BooleanField()
+    invit_icalandar = serializers.BooleanField(
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/serializers.py]: `invit_icalandar` -> `invit_icalendar` little typo

File: source/apiVolontaria/volunteer/serializers.py [hunk 2]
```diff
@@ -414,11 +415,39 @@ class ParticipationBasicSerializer(serializers.ModelSerializer):
...
     )
 
+    def create(self, validated_data):
+        participation = models.Participation()
+        participation.standby = validated_data['standby']
+        participation.event = validated_data['event']
+        participation.user = validated_data['user']
+
+        if 'invit_icalandar' in validated_data.keys():
+            participation.presence_duration_minutes = validated_data['invit_icalandar']
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/serializers.py]: Why did you change the duration minute of the participation in function of the fact we want an 'ics' file or not ? I'm lost here.

File: source/apiVolontaria/volunteer/serializers.py [hunk 3]
```diff
@@ -414,11 +415,39 @@ class ParticipationBasicSerializer(serializers.ModelSerializer):
...
+            name = "Volontaria event - " + participation.event.cycle.name
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/serializers.py]: I18N function need to be used here

File: source/apiVolontaria/volunteer/serializers.py [hunk 4]
```diff
@@ -414,11 +415,39 @@ class ParticipationBasicSerializer(serializers.ModelSerializer):
...
     user = UserBasicSerializer(
         read_only=True,
         default=serializers.CurrentUserDefault(),
     )
 
+    def create(self, validated_data):
+        participation = models.Participation()
+        participation.standby = validated_data['standby']
+        participation.event = validated_data['event']
+        participation.user = validated_data['user']
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/serializers.py]: It's working but it could be great to use this kind of logic to be more generic and help in the future :

```python
    def create(self, validated_data):
        invitation_ics = validated_data.pop('invit_icalendar')
        participation = models.Participation.objects.create(**validated_data)
        # Add here your logic for ICS file
        return participation
```

File: source/apiVolontaria/volunteer/serializers.py (hunk 1 above)
Comment (willkoua) [reply source/apiVolontaria/volunteer/serializers.py]: done

File: source/apiVolontaria/volunteer/serializers.py (hunk 2 above)
Comment (willkoua) [reply source/apiVolontaria/volunteer/serializers.py]: done

File: source/apiVolontaria/volunteer/serializers.py (hunk 3 above)
Comment (willkoua) [reply source/apiVolontaria/volunteer/serializers.py]: @RignonNoel je ne comprends pas pourquoi tu voudrais une traductions pour le nom de l'event

Comment (willkoua) [reply source/apiVolontaria/volunteer/serializers.py]: done

File: source/apiVolontaria/volunteer/serializers.py (hunk 4 above)
Comment (willkoua) [reply source/apiVolontaria/volunteer/serializers.py]: done

File: source/apiVolontaria/apiVolontaria/views.py [hunk 5]
```diff
@@ -340,7 +340,7 @@ def post(self, request, *args, **kwargs):
             }
 
             plain_msg = render_to_string("reset_password.txt", merge_data)
-            msg_html = render_to_string("reset_password.html", merge_data)
+            msg_html = render_to_string("calendar_invitation.html", merge_data)
```
Comment (RignonNoel) [inline source/apiVolontaria/apiVolontaria/views.py]: Hum.. What's the link with the fact we reset a password ?!

File: source/apiVolontaria/volunteer/serializers.py (hunk 4 above)
Comment (RignonNoel) [reply source/apiVolontaria/volunteer/serializers.py]: The second line of my example `participation = models.Participation.objects.create(**validated_data)` can replace all your lines:
```python
        participation = models.Participation()
        participation.standby = validated_data['standby']
        participation.event = validated_data['event']
        participation.user = validated_data['user']

        ....

        participation.save()
```

File: source/apiVolontaria/volunteer/templates/calendar_invitation.txt [hunk 6]
```diff
@@ -0,0 +1,16 @@
+Récupérer le calendrier de vos participations à nos évènements
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/templates/calendar_invitation.txt]: I18N in this file

File: source/apiVolontaria/volunteer/templates/rest_framework_docs/docs.html [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+{% block logo %}<a class="navbar-brand" href="#">API-Volontaria Documentation</a>{% endblock %}
+
+{% block jumbotron %}
+<div class="jumbotron">
+  <h1>Volontaria Web API</h1>
+  <h3>Documentation of the 'API-Volontaria' Web API.</h3>
+</div>
+{% endblock %}
+
+{% block footer %}<div class="footer">Copyright © 2017 Volontaria.</div>{% endblock %}
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/templates/rest_framework_docs/docs.html]: What's the purpose of this file please ?

File: source/apiVolontaria/volunteer/serializers.py [hunk 8]
```diff
@@ -414,11 +418,38 @@ class ParticipationBasicSerializer(serializers.ModelSerializer):
...
         read_only=True,
         default=serializers.CurrentUserDefault(),
     )
 
+    def create(self, validated_data):
+        invitation_ics = validated_data.pop('invit_icalendar')
+        participation = models.Participation.objects.create(**validated_data)
+
+        if invitation_ics:
+            if (os.path.exists(settings.MEDIA_ROOT) == False):
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/serializers.py]: You need to check if the `settings.CONSTANT['EMAIL_SERVICE']` is `True` or send an error because you can't send the ICS for the moment.

Actually i have an error 500 from the API if my email service is not configured

File: source/apiVolontaria/volunteer/templates/calendar_invitation.txt [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+--------------------------------------------------------------
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/templates/calendar_invitation.txt]: Add I18N to source/apiVolontaria/volunteer/templates/calendar_invitation.txt

(third time :p )

File: source/apiVolontaria/apiVolontaria/views.py [hunk 10]
```diff
@@ -154,7 +154,7 @@ def post(self, request, *args, **kwargs):
                 merge_data = {
                     'ACTIVATION_URL': FRONTEND_SETTINGS['ACTIVATION_URL'].replace(
                         "{{token}}",
-                        activate_token
+                        activation_url
```
Comment (RignonNoel) [inline source/apiVolontaria/apiVolontaria/views.py]: Hum, pourquoi touche tu a cette variable dans cette PR ?
Je suis d'accord qu'il y a un copier coller des lignes 145->148 qui est sale mais tu viens de briser la logique en changant cette variable.

File: source/apiVolontaria/volunteer/serializers.py [hunk 11]
```diff
@@ -429,11 +471,13 @@ class Meta:
             'subscription_date',
             'presence_duration_minutes',
             'presence_status',
+            'invit_icalendar',
         )
         read_only_fields = [
             'id',
             'presence_duration_minutes',
             'presence_status',
+            'invit_icalendar',
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/serializers.py]: invit_icalendar n'est pas read_only au contraire il est write_only comme tu l'a deja specifier a la ligne 424 de ce meme fichier

File: source/apiVolontaria/volunteer/serializers.py [hunk 12]
```diff
@@ -414,11 +418,48 @@ class ParticipationBasicSerializer(serializers.ModelSerializer):
...
+            if os.path.isfile(file_event_path):
+                merge_data = {
+                    'DOWNLOAD_URL': file_event_path,
+                    'CSS_STYLE': render_to_string('css/calendar_invitation.css')
+                }
+
+                plain_msg = render_to_string("calendar_invitation.txt")
+                msg_html = render_to_string("calendar_invitation.html", merge_data)
+                service_send_mail([participation.user.email], _('invitation calandar'), plain_msg, msg_html)
+            else:
```
Comment (RignonNoel) [inline source/apiVolontaria/volunteer/serializers.py]: The else is not on the good condition, did you try it ?
You will need to test it to be sure it work (add a test in test file).

**Comment by willkoua** [comment]
@RignonNoel il y a un test qui ne passe pas. DE mon coté, ce test fonctionne trés bien. Qu'est ce que je fais?
